File: benchmarks/pascal/mandelbrot.pas
program Mandelbrot;

{$mode objfpc}{$H+}

uses
  SysUtils, DateUtils;

const
  WIDTH = 800;
  HEIGHT = 800;
  MAX_ITERATIONS = 100;

type
  TResultArray = array of array of Integer;

{ Calculate the Mandelbrot set }
function CalculateMandelbrot(MaxIterations: Integer): TResultArray;
var
  x, y, iteration: Integer;
  cx, cy, zx, zy, temp: Double;
begin
  SetLength(Result, HEIGHT, WIDTH);
  
  for y := 0 to HEIGHT - 1 do
  begin
    for x := 0 to WIDTH - 1 do
    begin
      cx := (x - WIDTH / 2.0) * 4.0 / WIDTH;
      cy := (y - HEIGHT / 2.0) * 4.0 / HEIGHT;
      
      zx := 0.0;
      zy := 0.0;
      iteration := 0;
      
      while ((zx * zx + zy * zy <= 4.0) and (iteration < MaxIterations)) do
      begin
        temp := zx * zx - zy * zy + cx;
        zy := 2.0 * zx * zy + cy;
        zx := temp;
        Inc(iteration);
      end;
      
      Result[y, x] := iteration;
    end;
  end;
end;

{ Count non-black pixels in the result }
function CountNonBlack(const Result: TResultArray; MaxIterations: Integer): Integer;
var
  x, y: Integer;
begin
  CountNonBlack := 0;
  
  for y := 0 to HEIGHT - 1 do
  begin
    for x := 0 to WIDTH - 1 do
    begin
      if Result[y, x] < MaxIterations then
        Inc(CountNonBlack);
    end;
  end;
end;

var
  Result: TResultArray;
  Count: Integer;
  StartTime: TDateTime;
begin
  StartTime := Now;
  
  Result := CalculateMandelbrot(MAX_ITERATIONS);
  Count := CountNonBlack(Result, MAX_ITERATIONS);
  
  WriteLn('Mandelbrot set calculation finished.');
  WriteLn('Image size: ', WIDTH, ' x ', HEIGHT);
  WriteLn('Maximum iterations: ', MAX_ITERATIONS);
  WriteLn('Non-black pixels: ', Count);
  
  WriteLn('Time taken: ', MilliSecondsBetween(Now, StartTime), ' ms');
  
  // Approximate memory usage - this is a simplification
  WriteLn('Memory used: ', GetHeapStatus.TotalAllocated div 1024, ' KB');
end.

File: benchmarks/erlang/mandelbrot.erl
-module(mandelbrot).
-export([main/0]).

% Size constants
-define(WIDTH, 800).
-define(HEIGHT, 800).
-define(MAX_ITERATIONS, 100).

% Calculate the Mandelbrot set
calculate_mandelbrot(MaxIterations) ->
    lists:map(fun(Y) ->
        lists:map(fun(X) ->
            Cx = (X - ?WIDTH / 2.0) * 4.0 / ?WIDTH,
            Cy = (Y - ?HEIGHT / 2.0) * 4.0 / ?HEIGHT,
            calculate_pixel(Cx, Cy, MaxIterations)
        end, lists:seq(0, ?WIDTH - 1))
    end, lists:seq(0, ?HEIGHT - 1)).

calculate_pixel(Cx, Cy, MaxIterations) ->
    iterate(0.0, 0.0, Cx, Cy, 0, MaxIterations).

iterate(Zx, Zy, Cx, Cy, Iteration, MaxIterations) when Zx*Zx + Zy*Zy =< 4.0, Iteration < MaxIterations ->
    Temp = Zx*Zx - Zy*Zy + Cx,
    iterate(Temp, 2.0*Zx*Zy + Cy, Cx, Cy, Iteration + 1, MaxIterations);
iterate(_, _, _, _, Iteration, _) ->
    Iteration.

% Count non-black pixels in the result
count_non_black(Result, MaxIterations) ->
    lists:foldl(fun(Row, RowAcc) ->
        RowAcc + lists:foldl(fun(Pixel, PixelAcc) ->
            case Pixel < MaxIterations of
                true -> PixelAcc + 1;
                false -> PixelAcc
            end
        end, 0, Row)
    end, 0, Result).

main() ->
    StartTime = erlang:system_time(millisecond),
    
    Result = calculate_mandelbrot(?MAX_ITERATIONS),
    Count = count_non_black(Result, ?MAX_ITERATIONS),
    
    io:format("Mandelbrot set calculation finished.~n"),
    io:format("Image size: ~p x ~p~n", [?WIDTH, ?HEIGHT]),
    io:format("Maximum iterations: ~p~n", [?MAX_ITERATIONS]),
    io:format("Non-black pixels: ~p~n", [Count]),
    
    ElapsedTime = erlang:system_time(millisecond) - StartTime,
    io:format("Time taken: ~p ms~n", [ElapsedTime]),
    
    {ok, [TotalWords, _]} = process_info(self(), memory),
    io:format("Memory used: ~p KB~n", [TotalWords div 1024]),
    ok.